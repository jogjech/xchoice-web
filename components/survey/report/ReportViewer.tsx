import React, { FunctionComponent, useState, useEffect } from "react";
import { getSurveyReport } from "../../../apis/survey";
import { SurveyReport } from "../../../models/survey";
import QuestionReport from "./QuestionReport";
import { Alert, Statistic } from "antd";

interface Props {
  surveyId: string;
}

const ReportViewer: FunctionComponent<Props> = ({ surveyId }) => {
  const [surveyReport, setSurveyReport] = useState<SurveyReport>(undefined);
  const [loading, setLoading] = useState(true);
  const [errorMessage, setErrorMessage] = useState<string>(undefined);

  useEffect(() => {
    const fetchReport = async () => {
      const result = await getSurveyReport(surveyId);

      if (result.isError) {
        setErrorMessage(result.error.message);
      } else {
        setSurveyReport(result.surveyReport);
      }
      setLoading(false);
    };
    fetchReport();
  }, []);

  return (
    <>
      {loading ? (
        <div>loading...</div>
      ) : !!errorMessage ? (
        <Alert
          message="Failed to load response. Please refresh page to try again."
          type="error"
        ></Alert>
      ) : (
        <div>
          {surveyReport.published ? (
            <>
              <Statistic
                title="Responses captured"
                value={surveyReport.responses}
              ></Statistic>
              {surveyReport.responses === 0 ? (
                <div>No responses captured yet. Please check again later.</div>
              ) : (
                <div>
                  {surveyReport.questions.map((question, index) => (
                    <QuestionReport
                      question={question}
                      key={index}
                    ></QuestionReport>
                  ))}
                </div>
              )}
            </>
          ) : (
            <Alert
              message="Survey hasn't been published yet. Please check the response after publishing the survey."
              type="info"
              showIcon
            ></Alert>
          )}
        </div>
      )}
    </>
  );
};

export default ReportViewer;
