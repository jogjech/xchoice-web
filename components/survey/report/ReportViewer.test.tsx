import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import ReportViewer from "./ReportViewer";

describe("Report viewer", () => {
  let wrapper: ShallowWrapper;
  const surveyId = "surveyId";

  beforeEach(() => {
    wrapper = shallow(<ReportViewer surveyId={surveyId}></ReportViewer>);
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("renders loading status", () => {
    expect(wrapper.find("div").text()).toBe("loading...");
  });
});
