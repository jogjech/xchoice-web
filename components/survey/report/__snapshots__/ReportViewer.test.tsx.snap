// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report viewer renders correctly 1`] = `
<Fragment>
  <div>
    loading...
  </div>
</Fragment>
`;
